### ISSUE TITLE
cmd/compile: go:linkname prevents inlining

### ISSUE THREAD

**Issue by dvyukov** [golang/go#20019]
go version devel +33c3477039 Tue Apr 18 03:56:16 2017 +0000 linux/amd64

`sync_runtime_procPin` is marked with `//go:linkname sync_runtime_procPin sync.runtime_procPin`. Runtime marks it as inlinable (which is good). However, sync fails to inline it with `cannot inline runtime_canSpin: no function body`.
There are more functions exposed from runtime to sync, sync/atomic, reflect, time, os.
Can we inline them?

**Comment by dvyukov** [comment]
@randall77 @aclements @RLH @valyala

**Comment by valyala** [comment]
cc'ing @josharian

**Comment by josharian** [comment]
At first glance, looks like an import/export issue--the function body is not associated correctly (or perhaps not exported at all) when linkname is present.

cc @mdempsky @griesemer

**Comment by dvyukov** [comment]
It's possible to quickly estimate the benefits -- make procPin/Unpin public in runtime and directly call from sync.Pool. @valyala care to prototype?

**Comment by cherrymui** [comment]
@josharian If I understand correctly, linkname is resolved at link time (as the name suggests). The runtime package doesn't export those names.

**Comment by valyala** [comment]
> It's possible to quickly estimate the benefits -- make procPin/Unpin public in runtime and directly call from sync.Pool. @valyala care to prototype?

@dvyukov , sure:

```
name            old time/op  new time/op  delta
Pool-4          10.5ns ± 9%   8.8ns ± 9%  -15.59%  (p=0.000 n=10+10)
PoolOverflow-4  2.10µs ± 3%  1.95µs ± 2%   -7.29%  (p=0.000 n=10+10)
```

**Comment by josharian** [comment]
@cherrymui but I believe we do have inlining information in the export data for non-exported function, specifically for (multi-level) inlining.

**Comment by cherrymui** [comment]
@josharian you're right, thanks!
But the package that have functions implemented by another package may not import that package. `sync/atomic` doesn't import `runtime`, and the compiler doesn't even read `runtime.a` when compiling `sync/atomic`. For this to work, we'll need to add `import _ "runtime"` in `sync/atomic`?

**Comment by josharian** [comment]
@cherrymui true. And that doesn't seem too much to ask, if you're already using go:linkname. :) Although this also reproduces with package sync (in pool.go), which does already import package runtime.

**Comment by mdempsky** [comment]
I think the issue is just in bimport.go and trivially fixable. @dvyukov do you have an easy standalone test case for this?

**Comment by dvyukov** [comment]
@mdempsky no, I don't. I was looking at sync_runtime_procPin.

**Comment by mdempsky** [comment]
Thanks. I can repro the issue, but realized it's more involved than simply tweaking bimport.go.

Currently, we associate function inlining bodies only at the individual Node level. To fix this, we'll need to cross-reference them via LSyms.

**Comment by josharian** [comment]
Not sure that's worth it at the moment.

**Comment by CAFxX** [comment]
If this is fixed I believe we can also rewrite Get/Put so that the fast path (get from/put into l.private) is inlined.

**Comment by ugorji** [comment]
Gentle ping.

**Comment by mdempsky** [comment]
> Would it maybe be possible to put `ProcPin`/`ProcUnpin` in `internal`?

In general, my preference would be towards more internal packages, rather than more use of `//go:linkname`. So that has my support, if it's applicable for this particular problem. (I'd even argue that we should move most of the current runtime package to a new internal/runtime package so it can simply export anything needed by the rest of the standard library, and then the existing runtime package just contains the public APIs, implemented on top of the internal ones.)

As for enabling inlining of `//go:linkname`d functions, I still think that's doable, at least for "push"-style linknames (i.e., when a lower-level package like runtime explicitly declares a function to be made available in a higher-level package like sync or net). I don't have any immediate plans to tackle this though. I'd encourage trying to use inline packages first; and if that proves impractical, I'll reassess.
